src/System.Net.Utilities/src/System/Net/NetworkInformation/Ping.Unix.cs

```diff
@@ -2,33 +2,268 @@
+        private static readonly string s_discoveredPing4UtilityPath = GetPingUtilityPath(true);
+        private static readonly string s_discoveredPing6UtilityPath = GetPingUtilityPath(false);
+
+        private static bool CheckRawSocketPermissions()
+        {
+            try
+            {
+                Socket s = new Socket(AddressFamily.InterNetwork, SocketType.Raw, ProtocolType.Icmp);
+                return true;
+            }
+            catch
+            {
+                return false;
+            }
+        }
+
+        private static string GetPingUtilityPath(bool ipv4)
+        {
+            string fileName = ipv4 ? s_ipv4PingFile : s_ipv6PingFile;
+            foreach (string folder in s_binFolders)
+            {
+                string path = Path.Combine(folder, fileName);
+                if (File.Exists(path))
+                {
+                    return path;
+                }
+            }
+
+            return null;
+        }
+
+        private unsafe void InternalSendAsync(IPAddress address, byte[] buffer, int timeout, PingOptions options)
         {
             AsyncOperation asyncOp = _asyncOp;
             SendOrPostCallback callback = _onPingCompletedDelegate;
 
-            // TODO: Implement this (#2487)
+            PingReply pr = null;
+            Exception pingException = null;
+
+            try
+            {
+                if (s_canUseRawSockets)
+                {
+                    pr = SendIcmpPingOverRawSocket(address, buffer, timeout, options);
+                }
+                else
+                {
+                    pr = SendWithPingUtility(address, buffer, timeout, options);
+                }
+            }
+            catch (Exception e)
+            {
+                pingException = e;
+            }
 
-            PingException pe = new PingException(SR.net_ping, new PlatformNotSupportedException());
+            // At this point, either PR has a real PingReply in it, or pingException has an Exception in it.
             var ea = new PingCompletedEventArgs(
-                new PingReply(address, default(PingOptions), default(IPStatus), default(long), buffer), 
-                pe, 
-                false, 
+                pr,
+                pingException,
+                false,
                 asyncOp.UserSuppliedState);
 
             Finish();
             asyncOp.PostOperationCompleted(callback, ea);
         }
 
+        private unsafe PingReply SendIcmpPingOverRawSocket(IPAddress address, byte[] buffer, int timeout, PingOptions options)
+        {
+            EndPoint endPoint = new IPEndPoint(address, 0);
+
+            bool isIpv4 = address.AddressFamily == AddressFamily.InterNetwork;
+            ProtocolType protocolType = isIpv4 ? ProtocolType.Icmp : ProtocolType.IcmpV6;
+            // Use the current thread's ID as the identifier.
+            ushort identifier = (ushort)Environment.CurrentManagedThreadId;
```

Comment:
It is fairly arbitrary, it's just used to match the request we send with the response we receive back. This is apparently what the ping utility uses so I did the same.